### ISSUE TITLE
Question: hash password history?

### ISSUE THREAD

**Issue by rginnow** [imanghafoori1/laravel-password-history#1]
Sorry to raise an issue before there is actually any code, but how would one go about saving a history of hashed passwords? If I'm not mistaken, the password hash is based on the app key generated. If the app_key is ever changed wouldn't that fail any history checks?

**Comment by imanghafoori1** [comment]
Hi the simple answer is that the app key is not meant to be changed
Since if you change it, you  invalidate all the passwords and no one can
login

**Comment by rginnow** [comment]
Understandable, but couldn't you implement the "needsRehash" option which would force everyone to reset their password on the next login? If so, you couldn't rely on the password history to say something like "hey you've used that password before" at that point.

**Comment by imanghafoori1** [comment]
I think there is some thing here that I do not know about hashing.
Let's assume there is a very basic users table, with a password column and forget about pass history.

We have registered 10 users, and hashed their passwords. how do you want to validate their passwords when they want to login after you change the APP_KEY to a new one ?!

I think laravel auth()->check() will always return false, in this case.

2 - why we should change the app_key ? is the any compelling  reason ?

**Comment by rginnow** [comment]
My apologies, I incorrectly thought that the APP_KEY had something to do with the hashing, which apparently it does not :\

To me, since I thought the APP_KEY directly influenced the hashing algorithms, if the server was ever maliciously entered the malicious user could theoretically access the passwords. So I figured you'd have to change the APP_KEY and force password resets if ever a breach occurred.

I found a nice blog post on what it REALLY does.
https://tighten.co/blog/app-key-and-you

So... I guess a database of previously used passwords for users wouldn't be too bad as long as no one ends up finding a way of reversing the hash :)
